mesa_frames/abstract/space.py

```diff
@@ -0,0 +1,320 @@
+from abc import abstractmethod
+from collections.abc import Collection, Sequence
+from typing import TYPE_CHECKING
+
+from numpy.random import Generator
+from typing_extensions import Self
+
+from mesa_frames.abstract.agents import AgentContainer
+from mesa_frames.abstract.mixin import CopyMixin, DataFrameMixin
+from mesa_frames.types_ import (
+    DataFrame,
+    GeoDataFrame,
+    IdsLike,
+    SpaceCoordinate,
+    SpaceCoordinates,
+)
+
+ESPG = int
+
+if TYPE_CHECKING:
+    from mesa_frames.concrete.model import ModelDF
+
+
+class SpaceDF(CopyMixin, DataFrameMixin):
+    _model: "ModelDF"
+    _agents: DataFrame | GeoDataFrame
+
+    def __init__(self, model: "ModelDF") -> None:
+        """Create a new SpaceDF object.
+
+        Parameters
+        ----------
+        model : 'ModelDF'
+
+        Returns
+        -------
+        None
+        """
+        self._model = model
+
+    def random_agents(
+        self,
+        n: int,
+        seed: int | None = None,
+    ) -> DataFrame:
+        """Return a random sample of agents from the space.
+
+        Parameters
+        ----------
+        n : int
+            The number of agents to sample
+        seed : int | None, optional
+            The seed for the sampling, by default None
+            If None, an integer from the model's random number generator is used.
+
+        Returns
+        -------
+        DataFrame
+            A DataFrame with the sampled agents
+        """
+        if seed is None:
+            seed = self.random.integers(0)
+        return self._df_sample(self._agents, n=n, seed=seed)
+
+    @abstractmethod
+    def get_directions(
```

Comment:
I find it a bit confusing. Is direction a unit vector, or the entire direction vector which norm is > 1?

Reply:
I think it's fine to add a fine print saying it is non-normalized. It showed up in the discussion in mesa-examples in the boid flocker model as well. In the old Mesa space it is called `get_heading`, not that this is any clearer either. I have no strong opinion on the naming or maintaining compatibility for this particular method, because most models don't use it.

Reply:
I agree it won't be a particularly common method but having a normalize parameter with default false won't hurt. It's one line of code in the implementation anyway and I think it's clearer.

Reply:
OK SGTM.